### PR TITLE
:penguin: added zfs packages to all images

### LINKED ISSUE #510: 🐧 Zfs support
**Is your feature request related to a problem? Please describe.**

Zfs isn't supported currently as missing from core images https://github.com/kairos-io/kairos/tree/master/images

**Describe the solution you'd like**
Zfs to be shipped, so can be used for partitioning

**Additional context**
- [ ] Add zfs packages to flavors (check distro packages support, and be careful on how many deps introduces, in e.g. alpine)
- [ ] QA - can I create a zfs partition after install, and mount it over somewhere ? (vm with 2 disks)

### PR DESCRIPTION
**Pull request by paynejacob**
**What this PR does / why we need it**

This PR introduces zfs tools and kernel modules to all kairos images.  Additional test were added to ensure zfs kernel modules can be loaded and a zfs pools can be created on all os.  Booting from a zfs drive is not supported at this time.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #510

### CODE REVIEW

**Review by oz123 — COMMENTED** [review]
LGTM! The added test is really and useful for the custom mounts feature!

Thanks!

**Review by mudler — DISMISSED** [review]
Awesome, thanks!

**Comment by mudler** [comment]
mm alpine seems to fail reliably, just wondering if you need to omit zfs as well for it in dracut. If that's the case, needs also to omit it on kairos-io/packages, where the kernel and initrd for alpine are built

File: tests/zfs_test.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package mos_test
+
+import (
+	"fmt"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	. "github.com/spectrocloud/peg/matcher"
+)
+
+var _ = Describe("kairos zfs test", Label("zfs"), func() {
```
Comment (mudler) [inline tests/zfs_test.go:10]: plans to attach this test to the CI? I think we can give this at least for a flavor. or maybe put as part of our acceptance tests, as a separate block?

**Comment by paynejacob** [comment]
Had to remove the zfs packages from alpine-opensuse-leap, the packages were causing a kernel panic in the opensuse kernel.

File: tests/zfs_test.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		EventuallyConnects(1200)
+	})
+
+	Context("install", func() {
+		It("to disk with zfs kernel module loaded", func() {
+			err := Machine.SendFile("assets/zfs.yaml", "/tmp/zfs.yaml", "0770")
+			Expect(err).ToNot(HaveOccurred())
+
+			out, _ := Sudo("kairos-agent manual-install --device auto /tmp/zfs.yaml")
+			fmt.Println(out)
```
Comment (jimmykarily) [inline tests/zfs_test.go:21]:
```suggestion
```

File: tests/zfs_test.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+			_, _ = Sudo("sync")
```
Comment (jimmykarily) [inline tests/zfs_test.go:22]:
```suggestion
			out, err = Sudo("sync")
			Expect(err).ToNot(HaveOccurred(), out)
```

File: tests/zfs_test.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	BeforeEach(func() {
+		EventuallyConnects(1200)
+	})
+
+	Context("install", func() {
+		It("to disk with zfs kernel module loaded", func() {
+			err := Machine.SendFile("assets/zfs.yaml", "/tmp/zfs.yaml", "0770")
+			Expect(err).ToNot(HaveOccurred())
+
+			out, _ := Sudo("kairos-agent manual-install --device auto /tmp/zfs.yaml")
```
Comment (jimmykarily) [inline tests/zfs_test.go:20]:
```suggestion
			out,  err := Sudo("kairos-agent manual-install --device auto /tmp/zfs.yaml")
			Expect(err).ToNot(HaveOccurred())
```

File: tests/zfs_test.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	})
+
+	Context("create pool", func() {
+		It("with two disks and mount it at /mnt/pool0", func() {
+			_, err := Sudo("dd if=/dev/zero of=/usr/local/disk0.img bs=1M count=1024")
+			Expect(err).ToNot(HaveOccurred())
+			_, err = Sudo("dd if=/dev/zero of=/usr/local/disk1.img bs=1M count=1024")
+			Expect(err).ToNot(HaveOccurred())
+			out, err := Sudo("zpool create pool0 -m /usr/local/pool0 /usr/local/disk0.img /usr/local/disk1.img")
+			fmt.Println(out)
```
Comment (jimmykarily) [inline tests/zfs_test.go:34]:
```suggestion
			out, err := Sudo("zpool create pool0 -m /usr/local/pool0 /usr/local/disk0.img /usr/local/disk1.img")
			Expect(err).ToNot(HaveOccurred(), out)
```

File: tests/zfs_test.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("to be writable", func() {
+			out, err := Sudo("touch /usr/local/pool0/test")
+			fmt.Println(out)
+			Expect(err).ToNot(HaveOccurred())
```
Comment (jimmykarily) [inline tests/zfs_test.go:41]:
```suggestion
			Expect(err).ToNot(HaveOccurred(), out)
```

File: tests/zfs_test.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+		})
+
+		It("file to persist", func() {
+			out, err := Sudo("ls /usr/local/pool0")
+			fmt.Println(out)
+			Expect(err).ToNot(HaveOccurred())
```
Comment (jimmykarily) [inline tests/zfs_test.go:47]:
```suggestion
			Expect(err).ToNot(HaveOccurred(), out)
```

File: tests/zfs_test.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+			fmt.Println(out)
+			_, _ = Sudo("sync")
+			detachAndReboot()
+		})
+	})
+
+	Context("create pool", func() {
+		It("with two disks and mount it at /mnt/pool0", func() {
+			_, err := Sudo("dd if=/dev/zero of=/usr/local/disk0.img bs=1M count=1024")
+			Expect(err).ToNot(HaveOccurred())
```
Comment (jimmykarily) [inline tests/zfs_test.go:30]:
```suggestion
			out, err := Sudo("dd if=/dev/zero of=/usr/local/disk0.img bs=1M count=1024")
			Expect(err).ToNot(HaveOccurred(), out)
```

File: tests/zfs_test.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+			_, err = Sudo("dd if=/dev/zero of=/usr/local/disk1.img bs=1M count=1024")
+			Expect(err).ToNot(HaveOccurred())
```
Comment (jimmykarily) [inline tests/zfs_test.go:32]:
```suggestion
			out, err = Sudo("dd if=/dev/zero of=/usr/local/disk1.img bs=1M count=1024")
			Expect(err).ToNot(HaveOccurred(), out)
```

File: tests/zfs_test.go (hunk 1 above)
Comment (paynejacob) [reply tests/zfs_test.go:10]: I like the idea of running it for at least one flavor, I think acceptance tests make sense

File: .github/workflows/image.yaml [hunk 10]
```diff
@@ -151,6 +151,33 @@ jobs:
...
+  zfs-tests:
+    needs:
+      - build
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - flavor: "opensuse-leap"
+          - flavor: "opensuse-tumbleweed"
```
Comment (mudler) [inline .github/workflows/image.yaml:163]:
```suggestion
```

File: .github/workflows/image.yaml [hunk 11]
```diff
@@ -151,6 +151,33 @@ jobs:
...
+          - flavor: "ubuntu"
```
Comment (mudler) [inline .github/workflows/image.yaml:164]:
```suggestion
```

File: .github/workflows/image.yaml [hunk 12]
```diff
@@ -151,6 +151,33 @@ jobs:
...
+          - flavor: "debian"
```
Comment (mudler) [inline .github/workflows/image.yaml:165]:
```suggestion
```

File: .github/workflows/image.yaml [hunk 13]
```diff
@@ -151,6 +151,33 @@ jobs:
...
+          - flavor: "ubuntu-20-lts"
```
Comment (mudler) [inline .github/workflows/image.yaml:166]:
```suggestion
```

File: .github/workflows/image.yaml [hunk 14]
```diff
@@ -151,6 +151,33 @@ jobs:
...
+          - flavor: "ubuntu-22-lts"
```
Comment (mudler) [inline .github/workflows/image.yaml:167]:
```suggestion
```

File: .github/workflows/image.yaml (hunk 10 above)
Comment (mudler) [reply .github/workflows/image.yaml:163]: until we scale our CIs, better stick at one job only for PR/run

**Review by mudler — APPROVED** [review]
looking good here, just tests looks failing

**Review by jimmykarily — DISMISSED** [review]

File: overlay/files-opensuse/etc/dracut.conf.d/02-zfs.conf [hunk 15]
```diff
@@ -0,0 +1 @@
+omit_dracutmodules+=" zfs "
```
Comment (Itxaka) [inline overlay/files-opensuse/etc/dracut.conf.d/02-zfs.conf:1]: Why?

File: images/Dockerfile.opensuse-leap [hunk 16]
```diff
@@ -1,8 +1,16 @@
+# syntax=docker/dockerfile:1.4
 ARG BASE_IMAGE=opensuse/leap:15.4
 
 FROM $BASE_IMAGE
 
-RUN zypper in -y \
+ARG K3S_VERSION
```
Comment (Itxaka) [inline images/Dockerfile.opensuse-leap:6]: This was dropped already from all the docker files as it's not used anywhere:)

File: overlay/files-opensuse/etc/dracut.conf.d/02-zfs.conf (hunk 15 above)
Comment (Itxaka) [reply overlay/files-opensuse/etc/dracut.conf.d/02-zfs.conf:1]: oh wow, interesting, thanks for the info!

File: images/Dockerfile.opensuse-leap (hunk 16 above)
Comment (paynejacob) [reply images/Dockerfile.opensuse-leap]: ah that was probably just a bad rebase thanks!

**Review by mudler — APPROVED** [review]
Looks good!

**Comment by jimmykarily** [comment]
The problems on CI are caused by the recent refactoring we did to peg and ginkgo helper functions. @paynejacob since I had to fix other similar issues on other branches, want me to take over and make CI green? The rest of this PR looks good to merge.

**Comment by jimmykarily** [comment]
Opened a new issue for the trivy failure: https://github.com/kairos-io/kairos/issues/1065

@paynejacob this branch is not rebased and tests are green. Please have a look to see if anything is off after the rebase.

File: .github/workflows/image.yaml [hunk 17]
```diff
@@ -261,6 +261,40 @@ jobs:
...
+          # The default value is 'false'
+          latest: true
+          repository: "kairos-io/kairos"
+          fileName: "*opensuse-leap*.iso"
+          out-file-path: "last-release"
+      - uses: actions/upload-artifact@v3
+        with:
+          name: latest-release.zip
+          path: last-release
+          if-no-files-found: error
```
Comment (jimmykarily) [inline .github/workflows/image.yaml:276]: This is duplicate with the block above. Rebase "artifact" ?

File: Earthfile [hunk 18]
```diff
@@ -328,6 +328,23 @@ docker:
...
+    ELSE IF [ "$FLAVOR" = "alpine-arm-rpi" ]
+        COPY overlay/files-opensuse-arm-rpi/ /
+    ELSE IF [ "$FLAVOR" = "opensuse-leap-arm-rpi" ] || [ "$FLAVOR" = "opensuse-tumbleweed-arm-rpi" ]
+        COPY overlay/files-opensuse-arm-rpi/ /
+    ELSE IF [ "$FLAVOR" = "fedora" ] || [ "$FLAVOR" = "rockylinux" ]
+        COPY overlay/files-fedora/ /
+    ELSE IF [ "$FLAVOR" = "debian" ] || [ "$FLAVOR" = "ubuntu" ] || [ "$FLAVOR" = "ubuntu-20-lts" ] || [ "$FLAVOR" = "ubuntu-22-lts" ]
+        COPY overlay/files-ubuntu/ /
+    END
+
```
Comment (jimmykarily) [inline Earthfile:347]: How is this relevant to zfs? Maybe a rebase "artifact" too?
